### PR TITLE
Add UUID as a value type

### PR DESCRIPTION
**Pull request by markelliot**

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java [hunk 1]
```diff
@@ -840,6 +842,77 @@ public String getPersistCode(String variableName) {
...
+        }
+
+        @Override
+        public String getHydrateCode(String inputName, String indexName) {
+            return String.format("EncodingUtils.decodeUUID(%s, %s)", inputName, indexName);
+        }
+
+        @Override
+        public String getFlippedHydrateCode(String inputName, String indexName) {
+            return String.format("Long.MAX_VALUE ^ %s", getHydrateCode(inputName, indexName));
```
Comment (carrino) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java]: this is a bug I think

**Comment by carrino** [comment]
I think you'll have to add proto support for it also

**Comment by carrino** [comment]
basically everywhere you see FIXED_LONG_LITTLE_ENDIAN you'll have to push this through

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java (hunk 1 above)
Comment (markelliot) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java]: Er, yeah, I see, what's the right way to do this? EncodingUtils.flipAllBits?

Comment (carrino) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java]: EncodingUtils.decodeUUIDFlipped i think is the easiest

Comment (markelliot) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java]: okay, added EncodingUtils.decodeFlippedUUID; I think it's right now

File: atlasdb-client/src/main/java/com/palantir/atlasdb/ptobject/EncodingUtils.java [hunk 2]
```diff
@@ -255,6 +257,25 @@ public static int sizeOfSizedBytes(byte[] bytes) {
...
+    }
+
+    public static UUID decodeUUID(byte[] bytes, int offset) {
+        long mostSigBits = PtBytes.toLong(bytes, offset);
+        long leastSigBits = PtBytes.toLong(bytes, offset + 8);
+        return new UUID(mostSigBits, leastSigBits);
+    }
+
+    public static UUID decodeFlippedUUID(byte[] bytes, int offset) {
+        long mostSigBits = PtBytes.toLong(flipAllBitsInPlace(bytes, offset, 8), offset);
```
Comment (carrino) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/ptobject/EncodingUtils.java]: flip in place modifies the byte[].  I think just do PtBytes.toLong(bytes, offset) and PtBytes.toLong(bytes, offset + 8); and flip each with the standard ^ -1L trick.

Comment (markelliot) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/ptobject/EncodingUtils.java]: ok

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java [hunk 3]
```diff
@@ -840,6 +842,77 @@ public String getPersistCode(String variableName) {
             return Long.class;
         }
     },
+    UUID {
+        @Override
+        public UUID convertToJava(byte[] value, int offset) {
+            Preconditions.checkArgument(value.length == 16);
```
Comment (tjwilson90) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java]: This isn't necessarily true, the value could have stuff other than the uuid packed inside it.

**Comment by tjwilson90** [comment]
ValueType should probably get some tests.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java (hunk 3 above)
Comment (markelliot) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/ValueType.java]: ok, changed this to test that there's at least 16 more bytes available

**Comment by markelliot** [comment]
re: tests, yeah, I can make that part of this PR or we can do it as a follow-up. I could probably use some help gut-checking what things are actually supposed to do.

As a broader thing, I think we'd benefit from splitting schema definition from schema code-gen, and that doing so would likely make things easier to test.

**Comment by markelliot** [comment]
further thoughts here?

**Comment by markelliot** [comment]
@rjullman, rebased this, want to take a look?

**Comment by schlosna** [comment]
:+1: LGTM. I assume it is up to the client to use the proper version of UUID (i.e. version 4 or 5) to get uniform distribution to avoid unintentional data hot-spotting?

**Comment by markelliot** [comment]
Yep.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/encoding/PtBytes.java [hunk 4]
```diff
@@ -72,6 +75,16 @@ public static long toLong(byte[] bytes, int offset) {
     }
 
     /**
+     * Converts a UUID to a byte array.
+     */
+    public static byte[] toBytes(UUID id) {
```
Comment (markelliot) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/encoding/PtBytes.java]: note this is already in EncodingUtils

File: examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java [hunk 5]
```diff
@@ -49,22 +50,22 @@ public ProfileStore(TransactionManager txnMgr, Transaction t) {
...
     }
 
-    private long getNewId() {
+    private UUID getNewId() {
         long result = -1;
         while (result < 0) {
             result = new Random().nextLong();
         }
-        return result;
+        return new UUID(new Random().nextLong(), new Random().nextLong());
```
Comment (markelliot) [inline examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java]: see also UUID.randomUUID() :)

Comment (carrino) [inline examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java]: See also "all your entropy are belong to us".

File: examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java [hunk 6]
```diff
@@ -49,22 +50,18 @@ public ProfileStore(TransactionManager txnMgr, Transaction t) {
...
     }
 
-    private long getNewId() {
-        long result = -1;
-        while (result < 0) {
-            result = new Random().nextLong();
-        }
-        return result;
+    private UUID getNewId() {
+        return new UUID(new Random().nextLong(), new Random().nextLong());
```
Comment (rjullman) [inline examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java]: I'm happy to use UUID.randomUUID(), but heard that SecureRandom (used by the random uuid generation) has/causes entropy issues.

Comment (markelliot) [inline examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java]: Depends how frequently you're calling it. w/e, fine as is.

**Pull request merged**